File: java/divide_two_integers.java
public class Solution {
    public int divide(int dividend, int divisor) {
        if (divisor==0) {
            if (dividend>0) return Integer.MAX_VALUE;
            else return Integer.MIN_VALUE;
        }
        
        if (dividend==Integer.MIN_VALUE && divisor==-1) return Integer.MAX_VALUE;
        int res=0;
        boolean neg=(dividend>0&&divisor<0)||(dividend<0&&divisor>0);
        long dd=Math.abs((long)dividend);
        long dr=Math.abs((long)divisor);
        
        while (dd>=dr) {
            int i=0;
            while ((dr<<i+1)<dd) {
                i++;
            }
            res+=1<<i;
            dd-=(dr<<i);
        }
        
        if (neg) res=-res;
        return res;
    }
}

File: python/divide-two-integers.py
class Solution:
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        if divisor==0:
            return 2**31-1 if dividend!=0 else -2**31
        if dividend==-2**31 and divisor==-1:
            return 2**31-1
        
        res=0
        neg=(dividend>0) is not (divisor>0)        
        dd, dr=abs(dividend), abs(divisor)
        while (dd>=dr):
            i=0
            while (dd>=(dr<<i)): i+=1
            i-=1
            res+=1<<i
            dd-=(dr<<i)
        return res if neg==False else -res
